added roboflow example (#855)
neuralmagic/deepsparse: examples/roboflow-logging/README.md

# **Neural Magic + Roboflow**

> DeepSparse is more than the fastest deep learning inference runtime on CPUs - it also simplifies management of your production models. 

Reviewer: ```suggestion
DeepSparse is more than the fastest deep learning inference runtime on CPUs&mdashit also simplifies management of your production models. 
```

In order to combat these challenges, operations teams can adopt a **Continual Learning** mindset, where model performance is continually monitored, new 
data from production is continually gathered and labeled, and new versions of the model are continually trained. 

The workflow looks like this:

<p align="center">
    <img width=50% src="images/continual-learning.png">
</p>

Check out Full Stack Deep Learning's [awesome lecture](https://fullstackdeeplearning.com/course/2022/lecture-6-continual-learning/) 
for more details.

## **Tutorial Overview**

This example will demonstrate how to setup a Continual Learning system with Neural Magic and Roboflow
to train YOLOv5-s to detect soccer players in video clips.

There are 5 steps:
1. Deploy a YOLOv5-s model trained on COCO with DeepSparse
2. Log raw images from production to Roboflow
3. Label images with Roboflow
4. Retrain YOLOv5-s on the labeled production data with SparseML
5. Re-deploy the newly trained YOLOv5-s model with DeepSparse

Run the following to install the packages needed for the Tutorial. We recommend you

> use a virtual enviornment with Python.

Reviewer: ```suggestion
use a virtual environment with Python.
```

We can see DeepSparse achieves 241 items/second. ***This is an 5.8x performance gain over ORT!***

### **Deploying YOLOv5-s with DeepSparse**

Now that we have demonstrated the performance gains from running with DeepSparse, let's deploy a model.

DeepSparse offers both a Pipeline and Server API, enabling you to add DeepSparse to a Python application
or to spin up a model service endpoint. In this case, we will use DeepSparse Server.

#### **Launching DeepSparse Server**

DeepSparse Server is launched as a CLI command and is configured with a YAML file. The following YAML file
launches a object detection endpoint with DeepSparse running a pruned-quantized version of YOLOv5-s trained
on COCO (identified by the SparseZoo stub).

```yaml
## server-config-no-roboflow-logging.yaml
loggers: 
    python:

endpoints:
  - task: yolo
    model: zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned65_quant-none
    name: yolov5-s-coco
    route: /yolov5-s-coco/predict
```

Launch the Server. You should see Uvicorn report that it is running on `http://0.0.0.0:5543`.
```
deepsparse.server --config-file server/server-config-no-roboflow-logging.yaml
```

#### **Send Request to the Server**

> Now that the model endpoint is running, we can send raw images over HTTP and recieve back 

Reviewer: ```suggestion
Now that the model endpoint is running, we can send raw images over HTTP and receive back 
```

In our case, we will train with images of size 640x640, so leave all of the default options in place.

<p align="center">
  <img width="80%" src="images/upload-preprocessing.png">
</p>

Your dataset is ready for training!

### **Shortcut: Uploading Labeled Images From Roboflow Universe**

The dataset we are using has labels in the [Roboflow Universe](https://universe.roboflow.com/roboflow-jvuqo/football-players-detection-3zvbc/dataset/2), so we can skip the manual labeling for the purposes of this demo.

Download the data with labels in the YOLOv5 format:
```bash
mkdir data-with-labels; cd data-with-labels; curl -L "https://universe.roboflow.com/ds/YZbicMV8Z4?key=WrbJD7E7Ky" > roboflow.zip; unzip roboflow.zip; rm roboflow.zip; cd ..
```

Install the Roboflow CLI with `npm`.
```
npm i -g roboflow-cli
```

Authenticate to Roboflow (follow the prompts on the webpage).
```
roboflow auth
```

Import the dataset to your project.
```
roboflow import data-with-labels
```

Open your project on the Roboflow web app
```
roboflow open
```

Once you navigate to your project, you should see the 663 images! 

> Roboflow provides the option to perform pre-processing, augmentation, and generation to the dataset. In our case, we will train with images of size 640x640, so leave all of the default options in place.

Reviewer: ```suggestion
Roboflow provides the option to perform pre-processing, augmentation, and generation to the dataset. In our case, we will train with images of size 640x640; so, leave all of the default options in place.
```

<p align="center">
  <img width="80%" src="images/upload-preprocessing.png">
</p>

Your dataset is ready for training!

## **Step 4: Re-train the Model on Production Data**

In this step, we will use Neural Magic's SparseML and Roboflow's Datasets API to train a sparse version of YOLOv5 with the
data we gathered from production.

We have provided a [Colab Notebook](https://colab.research.google.com/drive/1e0zE6vGokh8LuXpDWhTzm-nelTHlrrFU)
with the full end-to-end training flow.

> ### **Download Data From Roboflow to Your Training Enviornment**

Reviewer: ```suggestion
### **Downloading Data from Roboflow to Your Training Environment**
```

> Roboflow has a convient CLI for downloading datasets.

Reviewer: ```suggestion
Roboflow has a convenient CLI for downloading datasets.
```

Make sure the CLI is installed.
```bash
npm i -g roboflow-cli
```

> Authenticate to Roboflow (follow the prompts on the webpage).

Reviewer: ```suggestion
Authenticate to Roboflow (follow the prompts on the web page).
```

```bash
roboflow auth
```

Download the dataset. Your $DATASET_URL should look something like `neural-magic/soccer-player-detection-gjrbw/2`.
```bash

> !roboflow download $DATASET_URL --format yolov5pytorch

Reviewer 2: ```suggestion
roboflow download $DATASET_URL --format yolov5pytorch
```

```

Create a new directory called `runs` inside the folder that was downloaded.
```
cd $DOWNLOADED_FOLDER
mkdir runs
cd runs
```

You are ready to train!

> ### **Train a Sparse YOLOv5 with SparseML**

Reviewer: ```suggestion
### **Training a Sparse YOLOv5 with SparseML**
```

> SparseML is an open-source optimization library which allows you to create sparse models trained on your dataset. One of the workflows enabled by 

Reviewer: ```suggestion
SparseML is an open-source optimization library that allows you to create sparse models trained on your dataset. One of the workflows enabled by 
```

> SparseML is called Sparse Transfer Learning, which is similiar to typical transfer learning, except

Reviewer: ```suggestion
SparseML is called Sparse Transfer Learning, which is similar to typical transfer learning, except
```

> you start from a sparse checkpoint and maintain the sparsity structure of the network while the fine-tuning occures.

Reviewer: ```suggestion
you start from a sparse checkpoint and maintain the sparsity structure of the network while fine-tuning occurs.
```

> SparseZoo contains sparse checkpoints for each version of YOLOv5 from N to X and SparseML is integrated with Ultralytics, so we can use a 

Reviewer: ```suggestion
SparseZoo contains sparse checkpoints for each version of YOLOv5 (from N to X) and SparseML is integrated with Ultralytics; so, we can use a 
```

familiar CLI script to launch training.

> #### **Kick Off Sparse Transfer Learning**

Reviewer: ```suggestion
#### **Kicking Off Sparse Transfer Learning**
```

Run the following to pull down the pruned-quantized YOLOv5s checkpoint and start fine-tuning. The `--recipe` argument points to a YAML file in SparseZoo that instruct SparseML to maintain sparsity levels as the training occurs.

```bash
sparseml.yolov5.train \
  --weights zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned65_quant-none?recipe_type=transfer_learn \
  --recipe zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned65_quant-none?recipe_type=transfer_learn \
  --cfg "yolov5s.yaml" \
  --data $DATASET_PATH/data.yaml \
  --cache 'ram'
```

> The training runs for 50 epochs, with final results by class looking like the following:

Reviewer: ```suggestion
The training runs for 50 epochs with final results by class looking like the following:
```

  all         38        905      0.969      0.221      0.273      0.142
    0         38         35          1          0          0          0
    1         38         27          1          0     0.0757     0.0325
    2         38        754      0.874      0.886      0.899      0.476
    3         38         89          1          0      0.117     0.0611
```

The model has a hard time picking up on classes 0, 1, and 3 (ball, goalkeeper, and referee), but does a nice job of identifying the players!

Pulling up the TensorBoard (`--logdir yolov5_runs/train`), we can see that sparsity levels (as seen in the `ParamPruning` group) remain flat for every layer of the model as the fine-tuning occured - SparseML is doing its job!

<p align="center">
  <img width="80%" src="images/param-pruning.png">
</p>

#### **Export to ONNX**

Run the following to convert the final PyTorch checkpoint to ONNX.
```
sparseml.yolov5.export_onnx --weights yolov5_runs/train/exp/weights/last.pt --dynamic
```

This creates an ONNX file representing the model at `yolov5_runs/train/exp/weights/last.onnx`. We will use 
this file when we redeploy the model with DeepSparse.

## **Step 5: Redeploy the New Model**

> We will create a new configuration file for the server which contains two endpoints - one with the sparse YOLOv5s

Reviewer: ```suggestion
We will create a new configuration file for the Server which contains two endpoints—one with the sparse YOLOv5s
```

trained on COCO and the sparse YOLOv5s fine-tuned on the production data.

Note that we provided a `last.onnx` file (which we trained in the manner described above) in the `server` directory.

> You can swap a path to your ONNX file if desired.

Reviewer: ```suggestion
You can swap a path to your ONNX file, if desired.
```

```yaml
# server-config-multiple-endpoints.yaml
loggers: 
    python:

endpoints:
  - task: yolo
    model: zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned65_quant-none
    name: yolov5-s-coco
    route: /yolov5-s-coco/predict
  
  - task: yolo
    model: server/last.onnx     # update with a path to your onnx file
    name: yolov5-s-finetuned
    route: /yolov5-s-finetuned/predict
```

Launch as before:
```
deepsparse.server --config-file server/server-config-multiple-endpoints.yaml
```

The following sends a request to each model endpoint and annotates the images with the bounding boxes, saving the output as files.

```
python client-multiple-endpoints.py
```

We can see that the fine-tuned model does a better job than the off-the-shelf COCO trained model!

YOLOv5s Trained on COCO        | YOLOv5s Finetuned on Production Data
:-----------------------------:|:------------------------------------:
![](images/annotated-coco.png) | ![](images/annotated-finetuned.png)

## **Wrapping Up**

You have successfully created a data flywheel for continual learning with DeepSparse and Roboflow!

> **Interested In Deploying DeepSparse To Production? [Start your 90 day free trial!](https://neuralmagic.com/deepsparse-free-trial/)**

Reviewer: ```suggestion
**Interested in deploying DeepSparse to production? [Start your 90 day free trial!](https://neuralmagic.com/deepsparse-free-trial/)**
```
